_posts/2022-06-20-dask-multi-cloud.md

```diff
@@ -0,0 +1,1198 @@
+
+
+```python
+def show_all_workers():
+    my_org().compute(workers='ecmwf-1-0')
+    my_org().compute(workers='ecmwf-1-1')
+    my_org().compute(workers='ecmwf-1-2')
+    my_org().compute(workers='ecmwf-1-3')
+    my_org().compute(workers='eumetsat-2-0')
+    my_org().compute(workers='eumetsat-2-1')
+    my_org().compute(workers='eumetsat-2-2')
+    my_org().compute(workers='eumetsat-2-3')
+    sleep(0.5)
+
+show_all_workers()
+```
+
+## Running on the Cluster
+
+Now that there is a Dask client in scope, calculations will be run on the cluster. We can define the tasks to be run
+
+
+```python
+predictions = xarray.open_dataset('/data/ecmwf/000490262cdd067721a34112963bcaa2b44860ab.nc').chunk('auto')
+observations = xarray.open_dataset('/data/eumetsat/ad-hoc/observations.nc').chunk('auto')    
+    
+averages = predictions.mean('realization')
+diff = averages.isel(height=10) - observations
+```
+
+But when we try to perform the calculation it fails
+
+
+```python
+with pytest.raises(FileNotFoundError) as excinfo:
+    show_all_workers()
+    diff.compute()    
+
+str(excinfo.value)
+```
+
+
+
+
+    "[Errno 2] No such file or directory: b'/data/eumetsat/ad-hoc/observations.nc'"
+
+
+
+It fails because the Dask scheduler has sent some of the tasks to read the data to workers running in EUMETSAT. They cannot see the data in ECMWF, and nor do we want them too, because reading all that data between data centres would be too slow.
+
+# Data-Proximate Computation
+
+Dask has the concept of [resources](https://distributed.dask.org/en/stable/resources.html). Tasks can be scheduled to run only where a resource (such as a GPU or amount of RAM) is available. We can [abuse this mechanism](https://dask.discourse.group/t/understanding-work-stealing/335/13) to pin tasks to a data centre, by treating the data centre as a resource.
+
+To do this, when we create the workers we mark them as having a `pool-ecmwf` or `pool-eumetsat` resource. Then when we want to create tasks that can only run in one data centre, we annotate them as requiring the appropriate resource
+
+
+```python
+with (dask.annotate(resources={'pool-ecmwf': 1})):
+    predictions.mean('realization').isel(height=10).compute()
+```
+
+A [special Python context manager](https://github.com/gjoseph92/dask-worker-pools) can hide the nastiness, allowing us to run a calculation only where the data is available
+
+
+```python
+with pool('ecmwf'):    
+    predictions.mean('realization').isel(height=10).compute()
+```
+
+Better still, we can load the data inside the context manager block and it will carry the task annonation with it
+
+
+```python
+with pool('ecmwf'):
+    predictions = xarray.open_dataset('/data/ecmwf/000490262cdd067721a34112963bcaa2b44860ab.nc').chunk('auto')
+```
+
+Another context manager in the library, `propagate_pools`, ensures that this resource pinning is respected
```

Comment:
It might be helpful to expand on why ensuring the resource pinning is necessary.